src/sentry/api/endpoints/source_map_debug.py

```diff
@@ -185,15 +170,130 @@ def _get_filename_and_path(self, exception, frame_idx):
         return filename, abs_path
 
-    def _find_matches(self, release_artifacts, unified_path):
-        full_matches = [artifact for artifact in release_artifacts if artifact.name == unified_path]
-        partial_matches = [
+    def _find_matches(self, release_artifacts, unified_path, filename, release, event):
+        full_matches = [
             artifact
             for artifact in release_artifacts
-            if artifact.name.endswith(unified_path.split("/")[-1])
+            if artifact.name == unified_path
+            and self._verify_dist_matches(release, event, artifact, filename)
         ]
+        partial_matches = self._find_partial_matches(unified_path, release_artifacts)
         return full_matches, partial_matches
 
+    def _find_partial_matches(self, unified_path, artifacts):
+        filename = unified_path.split("/")[-1]
+        filename_matches = [artifact for artifact in artifacts if artifact.name.endswith(filename)]
+        artifact_names = [artifact.name.split("/") for artifact in filename_matches]
+        while any(artifact_names):
+            for i in range(len(artifact_names)):
+                if unified_path.endswith("/".join(artifact_names[i])):
+                    return [artifacts[i]]
+                artifact_names[i] = artifact_names[i][1:]
+        return []
+
+    def _extract_release(self, event, project):
+        release_version = event.get_tag("sentry:release")
+
+        if not release_version:
+            raise Exception
+        return Release.objects.get(organization=project.organization, version=release_version)
+
+    def _verify_dist_matches(self, release, event, artifact, filename):
+        try:
+            if event.dist is None and artifact.dist_id is None:
+                return
+            dist = Distribution.objects.get(release=release, name=event.dist)
+        except Distribution.DoesNotExist:
+            raise Exception(
+                SourceMapProcessingIssue.DIST_MISMATCH,
+                {"eventDist": event.dist, "filename": filename},
+            )
+        if artifact.dist_id != dist.id:
+            raise Exception(
+                SourceMapProcessingIssue.DIST_MISMATCH,
+                {"eventDist": dist.id, "artifactDist": artifact.dist_id, "filename": filename},
+            )
+        return True
+
+    def _find_matching_artifact(
+        self, release_artifacts, urlparts, abs_path, filename, release, event
+    ):
+        unified_path = self._unify_url(urlparts)
+        full_matches, partial_matches = self._find_matches(
+            release_artifacts, unified_path, filename, release, event
+        )
+
+        artifact_names = [artifact.name for artifact in release_artifacts]
+
+        if len(full_matches) == 0:
+            if len(partial_matches) > 0:
+                partial_match = partial_matches[0]
+                url_prefix = self._find_url_prefix(filename, partial_match.name)
+
+                raise Exception(
```

Comment:
@roggenkemper I would make a new error called `SourceMapException` that subclasses `Exception` so we can catch that specifically in our error handler.